encoding/fuzz_test.go

```diff
@@ -0,0 +1,183 @@
+//go:build go1.18
+// +build go1.18
+
+package encoding_test
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"testing"
+
+	"github.com/segmentio/parquet-go/encoding"
+	"github.com/segmentio/parquet-go/encoding/bytestreamsplit"
+	"github.com/segmentio/parquet-go/encoding/delta"
+	"github.com/segmentio/parquet-go/encoding/plain"
+	"github.com/segmentio/parquet-go/encoding/rle"
+	"github.com/segmentio/parquet-go/format"
+)
+
+func fuzzEncoding(fuzz func(e encoding.Encoding)) {
+	for _, test := range [...]struct {
+		scenario string
+		encoding encoding.Encoding
+	}{
+		{
+			scenario: "PLAIN",
+			encoding: new(plain.Encoding),
+		},
+
+		{
+			scenario: "RLE",
+			encoding: new(rle.Encoding),
+		},
+
+		{
+			scenario: "PLAIN_DICTIONARY",
+			encoding: new(plain.DictionaryEncoding),
+		},
+
+		{
+			scenario: "RLE_DICTIONARY",
+			encoding: new(rle.DictionaryEncoding),
+		},
+
+		{
+			scenario: "DELTA_BINARY_PACKED",
+			encoding: new(delta.BinaryPackedEncoding),
+		},
+
+		{
+			scenario: "DELTA_LENGTH_BYTE_ARRAY",
+			encoding: new(delta.LengthByteArrayEncoding),
+		},
+
+		{
+			scenario: "DELTA_BYTE_ARRAY",
+			encoding: new(delta.ByteArrayEncoding),
+		},
+
+		{
+			scenario: "BYTE_STREAM_SPLIT",
+			encoding: new(bytestreamsplit.Encoding),
+		},
+	} {
+		fuzz(test.encoding)
+	}
+}
+
+func FuzzByteArrayEncoding(f *testing.F) {
+	f.Fuzz(func(t *testing.T, input []byte) {
+		fuzzEncoding(func(e encoding.Encoding) {
+			fuzzByteArrayEncoding(t, input, e)
+		})
+	})
+}
+
+func FuzzFixedLenByteArrayEncoding(f *testing.F) {
+	f.Fuzz(func(t *testing.T, size int, input []byte) {
+		fuzzEncoding(func(e encoding.Encoding) {
+			fuzzFixedLenByteArrayEncoding(t, size, input, e)
+		})
+	})
+}
+
+func fuzzFixedLenByteArrayEncoding(t *testing.T, size int, input []byte, e encoding.Encoding) {
+	if size < 0 || size >= 1<<60 {
+		return
+	}
+
+	if !e.CanEncode(format.FixedLenByteArray) {
+		return
+	}
+
+	t.Logf("encoding: %s", e.String())
+	t.Logf("fuzzing data - size=%d input=%v", size, input)
+
+	buf := new(bytes.Buffer)
+	enc := e.NewEncoder(buf)
+	dec := e.NewDecoder(buf)
+
+	defer dec.Reset(buf)
+	defer enc.Reset(buf)
+	defer buf.Reset()
+
+	tmp := make([]byte, size)
+
+	if err := enc.EncodeFixedLenByteArray(size, input); err != nil {
+		if errors.Is(err, encoding.ErrNotSupported) {
+			return
+		}
+
+		if errors.Is(err, encoding.ErrInvalidArguments) {
+			return
+		}
+		t.Fatal("encode: ", err)
+	}
+
+	for i := 0; i < (len(input) / size); i++ {
+		n, err := dec.DecodeFixedLenByteArray(size, tmp)
```

Comment:
I think we may get a lot more leverage out of the fuzzing code if we configure the test to decode the input rather than encoding it. 

The idea is to validate that the decoders can handle malformed payloads, and don't panic or error in unexpected ways when they are exposed to invalid input.

This would be the pattern we want to use:

```go
dec := e.NewDecoder(bytes.NewReader(input))
tmp := make([]byte, length)
err := dec.DecodeFixedLenByteArray(size, tmp)
```